File: 프로그래머스/LV2/코딩테스트 연습/월간 코드 챌린지 시즌1/삼각 달팽이/삼각 달팽이.py
def solution(n):
    answer = []
    map = [[0]*n for _ in range(n)]
    i, direction, r_start, c_start = 1, 0, 0, 0

    for l in range(n, 0, -1):
        if direction % 3 == 0:
            for r_start in range(l):
                map[r_start+r_start][c_start] = i
                i += 1
            c_start += 1
        elif direction % 3 == 1:
            for c_start in range(l):
                map[r_start+l][c_start+c_start] = i
                i += 1
            r_start += 1
        elif direction % 3 == 2:
            for di in range(1,l+1):
                map[r_start+l-di][c_start+l-di] = i
                i += 1
            r_start += 1
        direction += 1

    for r_start in map:
        for i in r_start:
            if i == 0:
                break
            answer.append(i)
    return answer

print(solution(4))

File: 프로그래머스/LV2/코딩테스트 연습/월간 코드 챌린지 시즌1/삼각 달팽이/삼각 달팽이.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int n) 
{
    vector<int> answer, temp;
    vector<vector<int>> map;
    int i = 1, direction = 0, r_start = 0, c_start = 0;
    for(int i = 0; i < n; i++)
    {
        temp.clear();
        for(int i = 0; i < n; i++)temp.push_back(0);
        map.push_back(temp);
    }

    for(int l = n; l >= 1; l--)
    {
        if(direction % 3 == 0)
        {
            for(int r = 0; r < l; r++)
            {
                map[r_start+r][c_start] = i++;
            }
            c_start++;
        }
        else if(direction % 3 == 1)
        {
            for(int c = 0; c < l; c++)
            {
                map[r_start+l][c_start+c] = i++;
            }
            r_start++;            
        }
        else if(direction % 3 == 2)
        {
            for(int di = 1; di <= l; di++)
            {
                map[r_start+l-di][c_start+l-di] = i++;
            }
            r_start++;
        }
        direction++;
    }

    for(vector<int> row : map)
    {
        for(int i : row)
        {
            if(i == 0)break;
            answer.push_back(i);
        }
    }
    return answer;
}